docs/proposals/component-registration.md

```diff
@@ -0,0 +1,294 @@
+
+To fix Endpoints Problem \#2, this proposal moves the specification (and status) of external service provider addresses to its own top-level namespaced API resource (provisionally called **ExternalServiceTarget**). Endpoints Controller will then continue to manage the creation/updates/deletion of Endpoints for internal services based on a selection of Pod instances, but will also do the same for external services based on a selection of ExternalServiceTargets. Endpoints then becomes a (conceptually) read-only API for the user.
+
+
+### API Changes
+
+1. Add new namespaced top level resource: ExternalServiceTarget
+  - Listable, like most other top-level resources
+  - Have both Spec and Status (filterable on both/either)
+  - Spec would have address (ip or host) and ports (optionally named)
+  - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
+  - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
+  - Similar to Endpoints, but more static. Defines the set of expected service instances, rather than the set of Ready service instances.
+  - Have to be manually added for non-component external Service instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
+2. Add a new field ServiceSpec.TargetSelector that selects a subset of ExternalServiceTargets
+  - Analogous to the ServiceSpec.Selector-Pod relationship
+3. Add a new ExternalServiceTarget Controller that handles probing ExternalServiceTargets
+  - Probe results update ExternalServiceTarget.Status.Conditions
+4. Modify Endpoints Controller to manage ExternalServiceTarget Endpoints
+  - Readiness from ExternalServiceTarget.Status.Conditions analogous to PodStatus.Conditions
+5. Modify components to self-register ExternalServiceTargets (one per component instance) instead of an Endpoint
+  - Group ExternalServiceTargets (component instances) with labels (e.g. "type=apiserver", "component=core")
+6. Register component Services (with ServiceSpec.TargetSelector) as part of cluster deployment
+  - Services of new components can also be added at runtime, before any instances exist. Each instance would self-register a ExternalServiceTargets.
+7. Modify Service to be support both internal and external targets in the same Service (Pods + ExternalServiceTargets)
+8. Deprecate `/componentsstatuses` in favor of requesting a filtered list of ExternalServiceTargets
+  - ExternalServiceTargets are external-only, but have generic conditions (liveness and readiness)
+  - Endpoints are external + internal, but only have readiness (for now)
+
```

Comment:
Further work (not absolutely necessary for this to be useful) would include having a way to roll up status for a service by looking at the liveness/readiness of the endpoints for that service.  Right now there is no way to easily ask "is this service ready".  However, reducing endpoint status to service status is not an obvious thing so there would probably be many different strategies.